src/DynamoDBMessageRepository/DynamoDBMessageRepository.php

```diff
@@ -0,0 +1,140 @@
+<?php
+
+namespace EventSauce\MessageRepository\DynamoDBMessageRepository;
+
+use Aws\DynamoDb\DynamoDbClient;
+use Aws\DynamoDb\Marshaler;
+use Aws\ResultPaginator;
+use EventSauce\EventSourcing\AggregateRootId;
+use EventSauce\EventSourcing\Message;
+use EventSauce\EventSourcing\MessageRepository;
+use EventSauce\EventSourcing\Serialization\MessageSerializer;
+use EventSauce\EventSourcing\Header;
+use EventSauce\EventSourcing\UnableToPersistMessages;
+use EventSauce\EventSourcing\UnableToRetrieveMessages;
+use Ramsey\Uuid\Uuid;
+use Generator;
+
+use Throwable;
+
+class DynamoDBMessageRepository implements MessageRepository
+{
+    protected DynamoDbClient $client;
+    protected MessageSerializer $serializer;
+    protected string $tableName;
+
+    public function __construct(DynamoDbClient $client, MessageSerializer $serializer, string $tableName)
+    {
+        $this->client = $client;
+        $this->serializer = $serializer;
+        $this->tableName = $tableName;
+    }
+
+    public function persist(Message ...$messages): void
+    {
+        if (count($messages) === 0) {
+            return;
+        }
+
+        $items = [];
+        $marshaler = new Marshaler();
+
+        foreach ($messages as $message) {
+            $payload = $this->serializer->serializeMessage($message);
+            $event = $payload['headers'][Header::EVENT_ID] = $payload['headers'][Header::EVENT_ID] ?? Uuid::uuid4()->toString();
+            $aggregateRootId = $payload['headers'][Header::AGGREGATE_ROOT_ID] ?? null;
+            $eventType = $payload['headers'][Header::EVENT_TYPE] ?? null;
+            $aggregateRootVersion = $payload['headers'][Header::AGGREGATE_ROOT_VERSION] ?? 0;
+            $timeOfRecording = $payload['headers'][Header::TIME_OF_RECORDING];
+
+            $item = [
+                'event' => $marshaler->marshalValue($event),
+                'aggregateRootId' => $marshaler->marshalValue($aggregateRootId),
+                'eventType' => $marshaler->marshalValue($eventType),
+                'aggregateRootVersion' => $marshaler->marshalValue($aggregateRootVersion),
+                'timeOfRecording' => $marshaler->marshalValue($timeOfRecording),
+                'payload' => $marshaler->marshalItem(['payload' => $payload])['payload']
+            ];
+
+            $items[] = [
+                'PutRequest' => [
+                    'Item' => $item
+                ]
+            ];
+        }
+
+        $batchRequest = ['RequestItems' => [ $this->tableName => $items ]];
+
+        try {
+            $this->client->batchWriteItem($batchRequest);
```

Comment:
You might want to implement your own batching here, since a call to `->batchWriteItem` can contain at most 25 mutations (puts / deletes).

Reply:
Also, the response can contain a number of unprocessed items, which will need to be retried.

https://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_BatchWriteItem.html#API_BatchWriteItem_ResponseSyntax

I am not familiar enough with EventSauce, but this might cause issues with the ordering of events (and no guarantee that all events will be stored at all). Given the fact that `->batchWriteItem` as a whole is not atomic, I'm strongly doubting if it's the right candidate for storing events. A sequence of `->putItem`s wrapped in a DynamoDB transaction could be a lot safer.

Reply:
This should be solved in my last commit

Reply:
I'm not a 100% sure, because I haven't worked with the `async-aws` client, but it seems like you're still not accounting for the fact that `->batchWriteItem()` will return an array of failed items. You will have to retry those until they succeed, or revert the successful inserts to maintain consistency. Since the latter is also impossible to do with 100% reliability/consistency, again I think `->batchWriteItem()` is not the right tool for the job. The only way I currently see to achieve consistency in storing the messages, is `->putItem()`s within a transaction.